pkg/blobstore/quorum/quorum_blob_access.go

```diff
@@ -0,0 +1,218 @@
+	writeQuorum int
+	generator   random.ThreadSafeGenerator
+}
+
+// NewQuorumBlobAccess creates a BlobAccess that applies operations to a subset
+// of storage backends, retrying on infrastructure errors.  Read and write quorum
+// size should be chosen so that they overlap by at least one backend.
+// Note: Data is not replicated again after the original write.
+func NewQuorumBlobAccess(backends []blobstore.BlobAccess, readQuorum, writeQuorum int) blobstore.BlobAccess {
+	quorumBlobAccessPrometheusMetrics.Do(func() {
+		prometheus.MustRegister(quorumBlobAccessFindMissingSynchronizations)
+	})
+
+	return &quorumBlobAccess{
+		backends:    backends,
+		readQuorum:  readQuorum,
+		writeQuorum: writeQuorum,
+		generator:   random.FastThreadSafeGenerator,
+	}
+}
+
+func (ba *quorumBlobAccess) shuffledBackends() []blobstore.BlobAccess {
+	backends := make([]blobstore.BlobAccess, len(ba.backends))
+	copy(backends, ba.backends)
+
+	ba.generator.Shuffle(len(backends), func(i, j int) {
+		backends[i], backends[j] = backends[j], backends[i]
+	})
+
+	return backends
+}
+
+type getQuorumErrorHandler struct {
+	remainingBackends []blobstore.BlobAccess
+	remainingReads    int
+	retry             func(blobstore.BlobAccess) buffer.Buffer
+}
+
+func (eh *getQuorumErrorHandler) tryNextBackendOrError(err error) (buffer.Buffer, error) {
+    if len(eh.remainingBackends) > 0 {
+        nextBackend := eh.remainingBackends[0]
+        eh.remainingBackends = eh.remainingBackends[1:]
+        return eh.retry(nextBackend), nil
+    }
+    return nil, err
+}
+
+func (eh *getQuorumErrorHandler) OnError(err error) (buffer.Buffer, error) {
+    fallbackErr := status.Error(codes.Unavailable, "Too many backends unavailable")
+    if util.IsInfrastructureError(err) {
+        // I/O error.  Try again on another backend.
+        return eh.tryNextBackendOrError(fallbackErr)
+
+    } else if status.Code(err) == codes.NotFound {
+        // Not found.  Try again on another backend - if we haven't seen enough yet.
+        if eh.remainingReads <= 1 {
+            // Observed sufficient NotFounds.  Return conclusive NotFound.
+            return nil, err
+        }
+        eh.remainingReads--
+
+        // Haven't been able to check enough backends.  Can't conclude not found.
+        return eh.tryNextBackendOrError(fallbackErr)
+    }
+
+	return nil, err
+}
+
+func (eh getQuorumErrorHandler) Done() {}
+
+func (ba *quorumBlobAccess) get(getter func(b blobstore.BlobAccess) buffer.Buffer) buffer.Buffer {
+	backends := ba.shuffledBackends()
+
+	backend := backends[0]
+	remainingBackends := backends[1:]
+
+	return buffer.WithErrorHandler(
+		getter(backend),
+		&getQuorumErrorHandler{
+			remainingBackends: remainingBackends,
+			remainingReads:    ba.readQuorum,
+			retry:             getter,
+		})
+}
+
+func (ba *quorumBlobAccess) Get(ctx context.Context, digest digest.Digest) buffer.Buffer {
+	return ba.get(func(b blobstore.BlobAccess) buffer.Buffer {
+		return b.Get(ctx, digest)
```

Comment:
How do you deal with servers being down in such a way that calls to Get() spend a lot of time hanging, e.g. waiting for a TCP connection to establish?